packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -147,33 +239,116 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
     internal
     override
     {
-        (address payer,) = abi.decode(context, (address, uint256));
-        uint256 ethActualCharge = relayHub.calculateCharge(gasUseWithoutPost + gasUsedByPost, relayData);
-        _depositProceedsToHub(ethActualCharge);
-        uint256 remainingTokenBalance = token.balanceOf(address(this));
-        require(token.transfer(payer, remainingTokenBalance), "failed refund");
-        emit TokensCharged(gasUseWithoutPost, gasUsedByPost, ethActualCharge);
-    }
-
-    function _depositProceedsToHub(uint256 ethActualCharge) private {
-        UniswapV3Helper.swapToEth(
-            address(weth),
-            address(token),
-            ethActualCharge,
-            uniswapPoolFee,
-            uniswap
-        );
-        relayHub.depositFor{value : address(this).balance}(address(this));
+        (address payer, uint256 priceQuote, uint256 tokenPreCharge) = abi.decode(context, (address, uint256, uint256));
+        IERC20Metadata token = _getTokenFromPaymasterData(relayData.paymasterData);
```

Comment:
yes, don't assume you have paymasterData in the postRelayedCall (since in AA there is only context)

Reply:
Fixed